OpenRA.Mods.Common/Widgets/WidgetUtils.cs

```diff
@@ -214,46 +215,78 @@ public static string FormatTimeSeconds(int seconds, bool leadingMinuteZero)
 			return $"{minutes:D}:{seconds % 60:D2}";
 		}
 
-		public static string WrapText(string text, int width, SpriteFont font)
+		static StringBuilder SplitLine(ReadOnlySpan<char> line, int width, SpriteFont font, StringBuilder sb)
 		{
-			var textSize = font.Measure(text);
-			if (textSize.X > width)
+			// Scan forwards until we find the last word that fits
+			// This guarantees a small bound on the amount of string we need to search before a linebreak
+			var start = 0;
+			var posPrevSpace = 0;
+			for (var i = 0; i < line.Length; ++i)
 			{
-				var lines = text.Split('\n').ToList();
-
-				for (var i = 0; i < lines.Count; i++)
+				if (line[i] == ' ')
 				{
-					var line = lines[i];
-					if (font.Measure(line).X <= width)
-						continue;
-
-					// Scan forwards until we find the last word that fits
-					// This guarantees a small bound on the amount of string we need to search before a linebreak
-					var start = 0;
-					while (true)
+					var length = i - start;
+					if (length > 0)
 					{
-						var spaceIndex = line.IndexOf(' ', start);
-						if (spaceIndex == -1)
-							break;
+						var fragment = line.Slice(start, length);
+						if (font.Measure(fragment).X > width)
+						{
+							sb.Append(line.Slice(start, posPrevSpace - start)).Append('\n');
+							start = posPrevSpace + 1;
+						}
+
+						posPrevSpace = i;
+					}
+				}
+			}
 
-						var fragmentWidth = font.Measure(line.Substring(0, spaceIndex)).X;
-						if (fragmentWidth > width)
-							break;
+			return sb.Append(line.Slice(start, line.Length - start));
+		}
 
-						start = spaceIndex + 1;
-					}
+		public static string WrapText(string text, int width, SpriteFont font)
+		{
+			var textSize = font.Measure(text);
+			if (textSize.X <= width)
+				return text;
+
+			// Set capacity to twice the length, randomly chosen.
```

Comment:
I don't understand this comment. It's not random?

Reply:
Last time i used a fixed value you asked if I could explain why i choose it. Will remove the comment.

Reply:
Oh, you meant arbitrarily chosen?

How about something like `// Preallocate extra capacity for inserting newlines`?